docs/qaoa/qaoa_maxcut.ipynb

```diff
@@ -0,0 +1,574 @@
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "IqUn4uv9_IVo"
+   },
+   "outputs": [],
+   "source": [
+    "def estimate_cost(graph, samples):\n",
+    "    \"\"\"Estimate the cost function of the QAOA on the given graph using the\n",
+    "    provided computational basis bitstrings.\"\"\"\n",
+    "    cost_value = 0.0\n",
+    "\n",
+    "    # Loop over edge pairs and compute contribution.\n",
+    "    for u, v, w in graph.edges(data=True):\n",
+    "        u_samples = samples[str(u)]\n",
+    "        v_samples = samples[str(v)]\n",
+    "\n",
+    "        # Determine if it was a +1 or -1 eigenvalue.\n",
+    "        u_signs = (-1) ** u_samples\n",
+    "        v_signs = (-1) ** v_samples\n",
+    "        term_signs = u_signs * v_signs\n",
+    "\n",
+    "        # Add scaled term to total cost.\n",
+    "        term_val = np.mean(term_signs) * w[\"weight\"]\n",
+    "        cost_value += term_val\n",
+    "\n",
+    "    return -cost_value"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "XLO0RRZarb_a"
+   },
+   "source": [
+    "Now we can sample from the `qaoa_circuit` and use `estimate_expectation` to calculate the expectation value of the cost function for the circuit. Below, we use arbitrary values for $\\alpha$ and $\\beta$."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "gZmW7NkBrl5Z"
+   },
+   "outputs": [],
+   "source": [
+    "alpha_value = np.pi / 4\n",
+    "beta_value = np.pi / 2\n",
+    "sim = cirq.Simulator()\n",
+    "\n",
+    "sample_results = sim.sample(\n",
+    "    qaoa_circuit, params={alpha: alpha_value, beta: beta_value}, repetitions=20_000\n",
+    ")\n",
+    "print(f\"Alpha = {round(alpha_value, 3)} Beta = {round(beta_value, 3)}\")\n",
+    "print(f\"Estimated cost: {estimate_cost(working_graph, sample_results)}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "rBmCr_DCsbtf"
+   },
+   "source": [
+    "## 4. Outer loop optimization\n",
+    "Now that we can compute the cost function, we want to find the optimal cost. There are lots of different techniques to choose optimal parameters for the `qaoa_circuit`. Since there are only two parameters here ($\\alpha$ and $\\beta$), we can keep things simple and sweep over incremental pairings using `np.linspace` and track the minimum value found along the way."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "id": "ma0pVZwSThQx"
+   },
+   "outputs": [],
+   "source": [
+    "# Set the grid size = number of points in the interval [0, 2π).\n",
+    "grid_size = 5\n",
+    "\n",
+    "exp_values = np.empty((grid_size, grid_size))\n",
+    "par_values = np.empty((grid_size, grid_size, 2))\n",
+    "\n",
+    "for i, alpha_value in enumerate(np.linspace(0, 2 * np.pi, grid_size)):\n",
+    "    for j, beta_value in enumerate(np.linspace(0, 2 * np.pi, grid_size)):\n",
```

Comment:
It seems like this should be a run_sweep instead of sampling on each parameter value individually.